### PR TITLE
File writer - interval

### PR DESCRIPTION
**Pull request by webbpinner**
Adds ability to truncate files every x hours or x minutes.  Before committing please ponder any negative performance impacts cause by the more complex processing of the file name... this is working for my test case but may be a problem for higher data rate streams.

### CODE REVIEW

File: logger/writers/file_writer.py [hunk 1]
```diff
@@ -38,6 +40,9 @@ def __init__(self, filename=None, mode='a', delimiter='\n', flush=True,
                      filename. By overridding time_format, other split
                      intervals, such as hourly or monthly, may be imposed.
 
+        split_interval Splits files based on a defined hour or minute time
```
Comment (davidpablocohn) [inline logger/writers/file_writer.py:43]: Expand text a little to make it clear that suffix must be either 'H' or 'M'

File: logger/writers/file_writer.py [hunk 2]
```diff
@@ -60,9 +65,55 @@ def __init__(self, filename=None, mode='a', delimiter='\n', flush=True,
...
         self.time_zone = time_zone
 
-        if split_by_time and not filename:
+        if (split_by_time or split_interval) is not None and not filename:
             raise ValueError('FileWriter: filename must be specified if '
-                             'split_by_time is True.')
+                             'split_by_time is specified or split_interval '
+                             'is True.')
+
+        if split_interval is None:
```
Comment (davidpablocohn) [inline logger/writers/file_writer.py:73]: Get rid of conditional, and have
```
self.split_interval = None
self.next_file_split = 0
```
We want self.split_interval to be defined, regardless.

Then see below for checking whether it's time to roll over to a new file.

File: logger/writers/file_writer.py [hunk 3]
```diff
@@ -60,9 +65,55 @@ def __init__(self, filename=None, mode='a', delimiter='\n', flush=True,
...
+            self.split_interval = None    
+
+        else:
```
Comment (davidpablocohn) [inline logger/writers/file_writer.py:76]: Break the parsing of the interval off into its own little method, for clarity.  Maybe _split_interval_in_seconds()?

File: logger/writers/file_writer.py [hunk 4]
```diff
@@ -120,7 +190,7 @@ def write(self, record):
 
         # If we're splitting by some time interval, see if it's time to
         # roll over to a new file.
-        if self.split_by_time:
+        if self.split_by_time or self.split_interval is not None:
```
Comment (davidpablocohn) [inline logger/writers/file_writer.py:193]: More efficient - if splitting by interval, we only call _get_new_file_suffix() when the right amount of time has passed.
```
    if self.split_by_time or (self.split_interval and time.now() > self.next_file_split):
      if self.split_interval:  # if we're splitting by interval, set time for our next split
        self.next_file_split = self.next_file_split + self.split_interval 
    new_file_suffix = ...  # rest of clause
        
```

**Review by davidpablocohn — CHANGES_REQUESTED** [review]
Can be more efficient with the interval checking by maintaining a "next_file_split" time variable - see inline.

File: logger/writers/file_writer.py (hunk 4 above)
Comment (webbpinner) [reply logger/writers/file_writer.py:193]: This is what I couldn't think of but figured was possible.  Makes much more sense.

**Comment by davidpablocohn** [comment]
Yeah, that's why I get the big bucks... :)

BTW: this PR makes me finally understand the request from SailDrone in the
survey about "different logging intervals." Seems like a fine idea - thanks
for implementing it!

**Comment by webbpinner** [comment]
Made changes based on recommendations.  These are untested but I wanted to make sure I was moving in the right direction.

**Review by davidpablocohn — APPROVED** [review]
Yup - this looks good.

**Comment by webbpinner** [comment]
ok... I'll actually test the code this next round and will report back.

**Comment by davidpablocohn** [comment]
Sounds good - thanks!

**Review by davidpablocohn — APPROVED** [review]

**Pull request merged**